driver-core/src/main/com/mongodb/ReadPreference.java

```diff
@@ -48,45 +48,6 @@
     ReadPreference() {
     }
 
-    /**
```

Comment:
I'm looking at this again, and realize that this change has to reverted.  I didn't realize that most of the factory methods in this class (e.g. `primaryPreferred()`) returns an instance of `ReadPreference` and not `TaggableReadPreference`, and we can't just change the return types of those methods as that would break binary compatibility (see 13.4.15 of https://docs.oracle.com/javase/specs/jls/se7/html/jls-13.html).

So even though this is a better API design (by preferring compile-time checks to runtime checks), it's not possible to do it safely: we have to move these methods back to `ReadPreference` and just document that they are not supported on primary read preference.

Sorry about the churn.

Reply:
Perhaps add a helpers  `ReadPreference.isTaggableReadPrefrence` and
 `ReadPreference.asTaggableReadPrefrence` to handle the cast.  Afaict only Primary isn't taggable.

Reply:
Reverted.